pennylane/plugins/default_qubit.py

```diff
@@ -218,10 +218,36 @@ def reset(self):
         self._state[0] = 1
         self._pre_rotated_state = self._state
 
-    def probability(self, wires=None):
+    def _analytic_probability(self, wires=None):
+        """Return the (marginal) analytic probability of each computational basis."""
         if self._state is None:
             return None
 
         wires = wires or range(self.num_wires)
+
         prob = self.marginal_prob(np.abs(self._state) ** 2, wires)
         return prob
+
+    def probability(self, wires=None):
+        """Return the (marginal) probability of each computational basis."""
+        wires = wires or range(self.num_wires)
+
+        if self.analytic:
+            return self._analytic_probability(wires=wires)
+
+        # non-analytic mode, estimate the probability from the generated samples
+
+        # consider only the requested wires
+        wires = np.hstack(wires)
+
+        samples = self._samples[:, np.array(wires)]
+
+        # convert samples from a list of 0, 1 integers, to base 10 representation
+        unraveled_indices = [2] * len(wires)
+        indices = np.ravel_multi_index(samples.T, unraveled_indices)
+
+        # count the basis state occurrences, and construct the probability vector
+        basis_states, counts = np.unique(indices, return_counts=True)
+        prob = np.zeros([len(wires) ** 2], dtype=np.float64)
+        prob[basis_states] = counts / self.shots
```

Comment:
Just for the clarity of the logic, it might be worth organizing this into a `_non_analytic_probability`. This would help us, in the long run, to review these parts of the codes more easily, and gives a clear view of how exactly the logic is branching.

The initial comment here, for example, would serve as a great one-liner in the docstring. :)

Reply:
Is this something that we would prefer as standard then, i.e. having three functions (`probability`, `_analytic_probability` and `non_analytic_probability`)?

Reply:
@josh146 what would be your take here?

From my side, I would be in favour of that. Just because this separation in logic would arise in any state vector simulator that is created for any plugin, given that the device supports both `analytic=True` and `analytic=False`.

Reply:
This is now the case I believe!